File: tools/og-card.ts
/**
 * Builds `public/og.png`, the 1200x630 link preview card (og:image / twitter:image).
 *
 *   npm run og:card
 *
 * The card is a real word cloud produced by the project's own code, not a drawing:
 * the palette comes from `src/theme` (`buildTheme` on one of the shipped `THEME_SPECS`),
 * the positions come from `layoutCloud` in `src/render/layout.ts`, and vertical CJK words
 * are stacked with the same `stackedLines` the canvas and the SVG export use. Changing a
 * theme therefore changes the card, and the card can never drift into a look the app
 * cannot produce.
 *
 * Text measurement and rasterisation go through `tools/og-card.swift` (CoreText, macOS
 * only) — see the header there for why. The output is committed, so this script runs only
 * when someone wants a new card; nothing in the build, the tests, the CI or the deploy
 * calls it.
 *
 * The word list below is invented product vocabulary. No chat log, character card or world
 * info has ever been near this file, and none may be added to it.
 */

import { spawnSync } from 'node:child_process';
import { mkdirSync } from 'node:fs';
import path from 'node:path';
import { buildTheme } from '../src/theme/palette';
import { THEME_SPECS } from '../src/theme/themes';
import { canStack, hashSeed, layoutCloud, stackedLines, type Measure } from '../src/render/layout';
import type { WordCount } from '../src/core/types';

const ROOT = path.resolve(import.meta.dirname, '..');
const OUT = path.join(ROOT, 'public', 'og.png');
const SWIFT = path.join(ROOT, 'tools', 'og-card.swift');

/** Open Graph's recommended size; X renders `summary_large_image` at this 1.91:1 ratio. */
const WIDTH = 1200;
const HEIGHT = 630;

/** Dark card: it sits well in both a light and a dark timeline, and matches `theme-color`. */
const THEME_ID = 'warm';

/**
 * What the tool is about, in both languages, with invented weights. `main` sorts by
 * weight before laying out — `layoutCloud` places in array order and never rotates the
 * first three, so the order has to be the frequency order or the card would put a minor
 * term in the middle.
 */
const WORDS: [string, number][] = [
  ['酒馆词云', 100], ['SillyTavern', 82], ['WordCloud', 74], ['聊天记录', 66],
  ['中文分词', 62], ['词频', 58], ['角色卡', 55], ['tokenize', 54], ['关键词', 52],
  ['offline', 50], ['世界书', 48], ['privacy', 47], ['离线', 46], ['隐私', 45],
  ['export', 44], ['浏览器', 43], ['导出', 42], ['chat log', 41], ['新词发现', 40],
  ['人名', 39], ['地点', 38], ['时间', 37], ['词类', 36], ['character card', 36],
  ['停用词', 35], ['清洗', 34], ['思维链', 33], ['world info', 33], ['状态栏', 32],
  ['群聊', 31], ['主题', 30], ['frequency', 30], ['PNG', 30], ['配色', 29],
  ['二维码', 28], ['SVG', 28], ['本地版', 27], ['entities', 27], ['单文件', 26],
  ['CSV', 26], ['词频表', 25], ['筛选', 24], ['open source', 24], ['zip', 24],
  ['同指', 23], ['别名', 22], ['jsonl', 22], ['无需词典', 21],
];

interface Sized { w: number; h: number }

/**
 * CoreText widths through `og-card.swift`. Measured once per string at `PROBE`, then
 * scaled: a CTLine's advance is linear in point size (no hinting at this path), so one
 * probe covers every font size the layout asks for. Verified below.
 */
const PROBE = 100;

function swift(mode: 'measure' | 'draw', payload: unknown): string {
  const r = spawnSync('swift', [SWIFT, mode], { input: JSON.stringify(payload), encoding: 'utf8', maxBuffer: 64 << 20 });
  if (r.error) throw new Error(`swift not runnable (${r.error.message}). This card can only be regenerated on macOS with the Xcode command line tools.`);
  if (r.status !== 0) throw new Error(`og-card.swift ${mode} failed:\n${r.stderr}`);
  return r.stdout;
}

/** CSS font-family string -> the family names, unquoted, generics dropped. */
function familyNames(stack: string): string[] {
  return stack
    .split(',')
    .map((s) => s.trim().replace(/^["']|["']$/g, ''))
    .filter((s) => s && !/^(sans-serif|serif|monospace|cursive|fantasy|system-ui|ui-\w+|-apple-system)$/.test(s));
}

function main(): void {
  const spec = THEME_SPECS.find((t) => t.id === THEME_ID);
  if (!spec) throw new Error(`no theme ${THEME_ID}`);
  const theme = buildTheme(spec);
  const family = familyNames(theme.cloudFont);
  const bold = true;

  // Every string the layout can ask about: whole words, plus each glyph of a word that
  // could end up as a vertical CJK column.
  const texts = new Set<string>();
  for (const [text] of WORDS) {
    texts.add(text);
    if (canStack(text)) for (const ch of text) texts.add(ch);
  }
  const measured = JSON.parse(swift('measure', { family, bold, size: PROBE, texts: [...texts] })) as Record<string, Sized>;

  // The linearity the single probe relies on, checked rather than assumed.
  const probe = [...texts][0];
  const half = JSON.parse(swift('measure', { family, bold, size: PROBE / 2, texts: [probe] })) as Record<string, Sized>;
  const ratio = half[probe].w / (measured[probe].w / 2);
  if (!(ratio > 0.99 && ratio < 1.01)) throw new Error(`CoreText width is not linear in font size (ratio ${ratio}); measure per size instead`);

  const measure: Measure = (text, fontSize) => {
    const m = measured[text];
    if (!m) throw new Error(`unmeasured string ${JSON.stringify(text)}`);
    const k = fontSize / PROBE;
    return { w: m.w * k, h: m.h * k };
  };

  const words: WordCount[] = [...WORDS].sort((a, b) => b[1] - a[1]).map(([text, count]) => ({ text, count }));
  // The bottom band carries the title and the URL, so the cloud is kept out of it.
  const inset = { top: 40, right: 48, bottom: 168, left: 48 };
  const placements = layoutCloud(words, {
    width: WIDTH,
    height: HEIGHT,
    maxFontSize: 62,
    minFontSize: 15,
    rotateRatio: 0.16,
    steps: theme.ramp.length,
    padding: 7,
    idleAmplitude: 0,
    seed: hashSeed('tavern-wordcloud/og'),
    fontFamily: theme.cloudFont,
    fontWeight: '600',
    inset,
  }, measure);

  const items: Record<string, unknown>[] = [];
  for (const p of placements) {
    const color = theme.ramp[Math.max(0, Math.min(theme.ramp.length - 1, p.step))];
    if (p.stacked) {
      for (const l of stackedLines(p.text, p.fontSize)) {
        items.push({ text: l.ch, x: p.x, y: p.y + l.dy, size: p.fontSize, color, bold });
      }
      continue;
    }
    // `rotate: 90` in CoreGraphics (y up) is the same quarter turn the SVG export writes
    // as rotate(-90) and the canvas renderer draws: the word reads bottom to top.
    items.push({ text: p.text, x: p.x, y: p.y, size: p.fontSize, color, bold, rotate: p.rotated ? 90 : 0 });
  }

  // Title band: name and URL on one baseline, the bilingual one-liner under them.
  const bandY = HEIGHT - inset.bottom + 88;
  const ink = theme.ramp[theme.ramp.length - 1];
  items.push({ text: '酒馆词云 · Tavern WordCloud', x: inset.left, y: bandY, size: 44, color: ink, bold, align: 'left', baseline: 'alphabetic' });
  items.push({ text: 'wordcloud.davidzhao.top', x: WIDTH - inset.right, y: bandY, size: 26, color: theme.accent, bold, align: 'right', baseline: 'alphabetic' });
  items.push({ text: '把 SillyTavern 聊天记录变成词云 · turn SillyTavern chat logs into a word cloud', x: inset.left, y: bandY + 42, size: 24, color: theme.fgDim, align: 'left', baseline: 'alphabetic' });

  mkdirSync(path.dirname(OUT), { recursive: true });
  swift('draw', { width: WIDTH, height: HEIGHT, background: theme.surface, family, out: OUT, items });
  console.log(`og.png  ${WIDTH}x${HEIGHT}  theme=${THEME_ID}  words=${placements.length}/${WORDS.length}  -> ${path.relative(ROOT, OUT)}`);
}

main();

File: tools/og-card.swift
// Text measurement and PNG painting for the social preview card (see tools/og-card.ts).
//
// Why a Swift helper: the card has to be a real raster PNG (X and Facebook do not render
// SVG in link previews) and it has to draw CJK. Nothing in the Node toolchain here can
// rasterise text — the repo has no canvas binding, and the screenshot tool is headless
// Chrome, which is far too heavy to sit in an asset pipeline. CoreText is already on every
// macOS box, needs no install and no network. It only ever runs when someone regenerates
// the card; `public/og.png` is committed, so a build, the CI and the deploy never touch it.
//
//   swift tools/og-card.swift measure < req.json > out.json
//   swift tools/og-card.swift draw    < spec.json          (writes spec.out)
//
// Both modes take the same font list and pick the first family the machine actually has,
// so a measured width is the width that later gets drawn.

import Foundation
import CoreGraphics
import CoreText
import ImageIO
import UniformTypeIdentifiers

func fail(_ msg: String) -> Never {
  FileHandle.standardError.write(("og-card.swift: " + msg + "\n").data(using: .utf8)!)
  exit(1)
}

func readStdinJSON() -> [String: Any] {
  let data = FileHandle.standardInput.readDataToEndOfFile()
  guard let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
    fail("stdin is not a JSON object")
  }
  return obj
}

/// First family in the list that this machine really has; the system font otherwise.
/// CTFontCreateWithName silently substitutes, so the resolved family name is compared back.
func resolveFamily(_ names: [String]) -> String? {
  for name in names {
    let f = CTFontCreateWithName(name as CFString, 12, nil)
    let got = CTFontCopyFamilyName(f) as String
    if got.caseInsensitiveCompare(name) == .orderedSame { return name }
  }
  return nil
}

func makeFont(_ family: String?, _ size: Double, bold: Bool) -> CTFont {
  let base: CTFont = family.map { CTFontCreateWithName($0 as CFString, size, nil) }
    ?? CTFontCreateUIFontForLanguage(.system, size, nil)!
  guard bold else { return base }
  return CTFontCreateCopyWithSymbolicTraits(base, size, nil, .traitBold, .traitBold) ?? base
}

func line(_ text: String, _ font: CTFont, _ color: CGColor) -> CTLine {
  let attrs: [CFString: Any] = [kCTFontAttributeName: font, kCTForegroundColorAttributeName: color]
  let astr = CFAttributedStringCreate(nil, text as CFString, attrs as CFDictionary)!
  return CTLineCreateWithAttributedString(astr)
}

/// `#rrggbb` (or `#rgb`) to a device-RGB colour. Anything else is a hard error: a silently
/// black word on a black card would be very easy to miss.
func parseColor(_ hex: String) -> CGColor {
  var s = hex.trimmingCharacters(in: .whitespaces)
  if s.hasPrefix("#") { s.removeFirst() }
  if s.count == 3 { s = s.map { "\($0)\($0)" }.joined() }
  guard s.count == 6, let v = UInt32(s, radix: 16) else { fail("bad colour \(hex)") }
  return CGColor(red: CGFloat((v >> 16) & 0xff) / 255, green: CGFloat((v >> 8) & 0xff) / 255,
                 blue: CGFloat(v & 0xff) / 255, alpha: 1)
}

let black = CGColor(red: 0, green: 0, blue: 0, alpha: 1)

// ── measure ──────────────────────────────────────────────────────────────────
// { family: [..], bold: Bool, size: Double, texts: [..] }
//   -> { "词": { "w": …, "h": … }, … }
// `w` is the advance width (what canvas measureText returns) and `h` the ink height
// (canvasMeasure's actualBoundingBoxAscent + Descent), so src/render/layout.ts gets
// exactly the two numbers its `Measure` contract asks for.
func runMeasure(_ req: [String: Any]) {
  let names = req["family"] as? [String] ?? []
  let bold = req["bold"] as? Bool ?? false
  let size = req["size"] as? Double ?? 100
  let texts = req["texts"] as? [String] ?? []
  let font = makeFont(resolveFamily(names), size, bold: bold)

  var out: [String: [String: Double]] = [:]
  for text in texts {
    let l = line(text, font, black)
    var asc: CGFloat = 0, desc: CGFloat = 0, lead: CGFloat = 0
    let advance = CTLineGetTypographicBounds(l, &asc, &desc, &lead)
    let ink = CTLineGetBoundsWithOptions(l, .useOpticalBounds)
    let h = ink.height > 0 ? Double(ink.height) : size * 0.92
    out[text] = ["w": Double(advance), "h": h]
  }
  let data = try! JSONSerialization.data(withJSONObject: out)
  FileHandle.standardOutput.write(data)
}

// ── draw ─────────────────────────────────────────────────────────────────────
// { width, height, background, family: [..], out: "path.png", items: [
//     { text, x, y, size, color, bold?, rotate?, align?: left|center|right,
//       baseline?: middle|alphabetic, opacity? } ] }
// Coordinates are top-left origin, like the layout and the SVG export; the context is
// native CoreGraphics (y up), so every y is flipped once here and nowhere else.
func runDraw(_ spec: [String: Any]) {
  let w = spec["width"] as? Int ?? 1200
  let h = spec["height"] as? Int ?? 630
  guard let outPath = spec["out"] as? String else { fail("draw needs `out`") }
  let names = spec["family"] as? [String] ?? []
  let family = resolveFamily(names)
  guard let items = spec["items"] as? [[String: Any]] else { fail("draw needs `items`") }

  let cs = CGColorSpaceCreateDeviceRGB()
  guard let ctx = CGContext(data: nil, width: w, height: h, bitsPerComponent: 8, bytesPerRow: 0,
                            space: cs, bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue) else {
    fail("could not create a \(w)x\(h) bitmap")
  }
  ctx.setFillColor(parseColor(spec["background"] as? String ?? "#000000"))
  ctx.fill(CGRect(x: 0, y: 0, width: w, height: h))
  ctx.setAllowsAntialiasing(true)
  ctx.setShouldSmoothFonts(true)

  for item in items {
    guard let text = item["text"] as? String, !text.isEmpty else { continue }
    let size = item["size"] as? Double ?? 16
    let bold = item["bold"] as? Bool ?? false
    let color = parseColor(item["color"] as? String ?? "#ffffff")
    let x = item["x"] as? Double ?? 0
    let y = Double(h) - (item["y"] as? Double ?? 0)      // top-left origin -> CG
    let rotate = item["rotate"] as? Double ?? 0           // degrees, counter-clockwise
    let align = item["align"] as? String ?? "center"
    let baseline = item["baseline"] as? String ?? "middle"
    let opacity = item["opacity"] as? Double ?? 1

    let l = line(text, makeFont(family, size, bold: bold), color)
    var asc: CGFloat = 0, desc: CGFloat = 0, lead: CGFloat = 0
    let advance = CTLineGetTypographicBounds(l, &asc, &desc, &lead)
    let ink = CTLineGetBoundsWithOptions(l, .useOpticalBounds)

    // Offset from the anchor to the text origin (the baseline's left end).
    var dx: Double
    switch align {
    case "left": dx = 0
    case "right": dx = -Double(advance)
    default: dx = -Double(advance) / 2
    }
    // "middle" centres the ink box on the anchor, which is what the canvas renderer and
    // the SVG export do for cloud words; "alphabetic" puts the baseline on it.
    let dy: Double = baseline == "alphabetic" ? 0 : -Double(ink.origin.y + ink.height / 2)

    ctx.saveGState()
    ctx.setAlpha(CGFloat(opacity))
    ctx.translateBy(x: CGFloat(x), y: CGFloat(y))
    if rotate != 0 { ctx.rotate(by: CGFloat(rotate * .pi / 180)) }
    ctx.textPosition = CGPoint(x: dx, y: dy)
    CTLineDraw(l, ctx)
    ctx.restoreGState()
  }

  guard let img = ctx.makeImage() else { fail("makeImage failed") }
  let url = URL(fileURLWithPath: outPath) as CFURL
  guard let dest = CGImageDestinationCreateWithURL(url, UTType.png.identifier as CFString, 1, nil) else {
    fail("cannot write \(outPath)")
  }
  CGImageDestinationAddImage(dest, img, nil)
  guard CGImageDestinationFinalize(dest) else { fail("PNG encoding failed") }
}

let mode = CommandLine.arguments.count > 1 ? CommandLine.arguments[1] : ""
switch mode {
case "measure": runMeasure(readStdinJSON())
case "draw": runDraw(readStdinJSON())
default: fail("usage: og-card.swift measure|draw  (JSON on stdin)")
}
